mythril/src/vm.rs

```diff
@@ -309,11 +317,27 @@ impl VirtualMachine {
     ) -> Result<Arc<RwLock<Self>>> {
         let guest_space = Self::setup_ept(&config, info)?;
 
-        Ok(Arc::new(RwLock::new(Self {
+        let vm = Arc::new(RwLock::new(Self {
             id: id,
             config: config,
             guest_space: guest_space,
-        })))
+            apic_access_page: Raw4kPage([0u8; 4096]),
+        }));
+
+        // Map the guest local apic addr to the access page. This will be set in each
+        // core's vmcs
+        let apic_frame = memory::HostPhysFrame::from_start_address(
+            memory::HostPhysAddr::new(
+                vm.read().apic_access_page.as_ptr() as u64
+            ),
+        )?;
+        vm.write().guest_space.map_frame(
```

Comment:
Why do we do this instead of handle this in the `EmulatedDevice`? I'm sure there is a reason, I'm just not following at the moment.

Reply:
Not sure what you're proposing. We have to map the `apic_frame` in to the guest address space at the `GUEST_LOCAL_APIC_ADDR`. This only happens once, as each vcpu uses the same `ApicAccessAddr` in their vmcs. The mapping here is essentially just necessary because according to the spec, if a memory access causes a VMEXIT due to an ept violation, it doesn't matter that the address was within the guest lapic range. EPT violations have a higher precedence. Therefore, we need to create a mapping to there _won't_ be an EPT violation. That allows the Apic Access violation to be the vmexit cause. That way, each vcpu can handle the access violation with their own local apic.

Reply:
Awesome! Thanks for the detailed explanation... that explains a lot now